Autonomous/Drive Control State Machine

StephenWelch opened the pull request:
Allows the driver to take control of the robot during autonomous and run commands during teleop while a button is held. Also includes fixes required to make unit testing work.

StephenWelch committed:
Start autonomous/driver control state machine

StephenWelch committed:
Merge branch 'master' into experimental/auton-statemachine

StephenWelch committed:
Merge remote-tracking branch 'origin/master' into experimental/auton-statemachine

StephenWelch committed:
Implement unit tests for DriverInput and make fixes

StephenWelch committed:
Update twice to verify that command triggering works as intended

StephenWelch committed:
Use Superstructure for autonomous commands

StephenWelch committed:
Remove extraneous prints/logs

StephenWelch committed:
Only use desktop NT for tests

StephenWelch committed:
Print stacktrace on Travis CI build

StephenWelch committed:
Print full exception on test failure

StephenWelch committed:
Possible workaround to Travis CI failure

StephenWelch committed:
Clean build and update

StephenWelch committed:
Use Ubuntu Xenial on Travis CI

StephenWelch committed:
Install specific lib needed

StephenWelch committed:
Use sudo

StephenWelch committed:
Install more things!

StephenWelch committed:
Whoops

StephenWelch committed:
Don't upgrade

StephenWelch committed:
Just exclude unit tests

oconnelc commented on robot/src/main/java/us/ilite/lib/drivers/Clock.java:
```diff
@@ -14,17 +14,22 @@
     private boolean hasTimeUpdatedThisCycle = false;
     private boolean mIsSimulated = false;
 
+    public Clock() {
```
Don't need this constructor

oconnelc commented on robot/src/main/java/us/ilite/lib/drivers/Clock.java:
```diff
@@ -41,15 +46,25 @@ public void cycleEnded() {
         return this;
     }
 
+    private double getJavaTime() {
```
Recommend making this a static method

oconnelc commented on robot/src/main/java/us/ilite/lib/drivers/Clock.java:
```diff
@@ -41,15 +46,25 @@ public void cycleEnded() {
+        return System.currentTimeMillis() / 1000.0;
+    }
+
+    private double getRobotTime() {
```
Recommend making this a static method

oconnelc commented on robot/src/main/java/us/ilite/robot/Robot.java:
```diff
@@ -91,26 +91,24 @@ public void autonomousInit() {
         mLogger.info("Starting Autonomous Initialization...");
 
+        mSuperstructure.startCommands(new FollowTrajectory(trajectory, mDrive, true));
+//        mCommandQueue.startCommands(new CharacterizeDrive(mDrive, false, false));
```
Can you take this line out?

oconnelc commented on robot/src/main/java/us/ilite/robot/commands/Delay.java:
```diff
@@ -1,10 +1,14 @@
+import com.flybotix.hfr.util.log.Logger;
+
 public class Delay implements ICommand {
+	private ILog mLog = Logger.createLog(Delay.class);
```
Recommend making this static

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -11,46 +13,51 @@
 public class DriverInput extends Module {
 
     private static final double DRIVER_SUB_WARP_AXIS_THRESHOLD = 0.5;
+    private ILog mLog = Logger.createLog(DriverInput.class);
```
Recommend making this static

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -11,46 +13,51 @@
-        this.mDriverJoystick = new Joystick(0);
-        this.mOperatorJoystick = new Joystick(1);
+        if(pSimulated) {
+            // Use a different joystick library?
```
Are things going to break if the mDriverJoystick and mOperatorJoystick are null? Maybe use Optional?

flybotix replied:
They aren't null; they are codexes instantiated at robot init.

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
+            rotate = Util.limit(rotate, 0.7);
+
+            if (mData.driverinput.get(DriveTeamInputMap.DRIVER_SUB_WARP_AXIS) > DRIVER_SUB_WARP_AXIS_THRESHOLD) {
+                throttle *= SystemSettings.kSnailModePercentThrottleReduction;
```
Will throttle ever get too big or too small with this logic?

flybotix replied:
Technically, throttle can be oversaturated.  Oversaturation is handled in the SRX if we don't handle it ourselves.

The throttle reduction constant reduces power.  Oversaturation isn't an issue here.

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
+
+        ETrackingType trackingType = null;
+        // Switch the limelight to a pipeline and track
+        if(mDriverInputCodex.isSet(DriveTeamInputMap.DRIVER_TRACK_TARGET_BTN)) {
```
Is it possible to have multiple targets selected? If so we need to handle this better.

flybotix replied:
I created a Trello card to handle this scenario.

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
+        boolean runCommands = false;
+        for(ELogitech310 l : SystemSettings.kTeleopCommandTriggers) {
+            if(mDriverInputCodex.isSet(l)) {
+                runCommands = true;
```
Recommend breaking

oconnelc commented on robot/src/main/java/us/ilite/robot/driverinput/DriverInput.java:
```diff
@@ -61,87 +68,108 @@ public void periodicInput(double pNow) {
+        boolean autonOverride = false;
+        for(ELogitech310 l : SystemSettings.kAutonOverrideTriggers) {
+            if(mDriverInputCodex.isSet(l) && mDriverInputCodex.get(l) > SystemSettings.kAutonOverrideAxisThreshold) {
+                autonOverride = true;
```
Recommend breaking

oconnelc commented on robot/src/test/java/us/ilite/robot/driverinput/DriverInputTest.java:
```diff
@@ -0,0 +1,141 @@
+public class DriverInputTest {
+
+    // We're only testing integration between Superstructure and DriverInput, so we can mock this
+    @Mock private Drive mDrive;
```
One problem with the mock annotations is that once the object is created, it's never re-created on subsequent tests. You'll want to have an "@after" annotation to null out the member,.

oconnelc reviewed (changes requested):
I have a couple of comments that should be looked at. Nothing too critical, but just some good housekeeping.

oconnelc committed:
Updated the Data class to use lazy loading for all of the tables so that the static values are only initialized when needed

oconnelc committed:
Moved the instantiation of the NetworkTableInstance

The instance is now being instantiated when it's used so that code
that doesn't want it, such as unit test doesn't

oconnelc committed:
Revert "Moved the instantiation of the NetworkTableInstance"

This reverts commit cd7e87f99a87bf2e79e1b7ae0fb3084c8b00ef7c.

oconnelc committed:
Revert "Updated the Data class to use lazy loading for all of the tables so that the static values are only initialized when needed"

This reverts commit 99d59a6366088154acdb56b8f6abde28ad0a826b.

flybotix committed:
Update Clock.java
